Ability to override OpenSSL version w/pkg_config

Why don't you make [use][2] of [pkg_config][1]?

[1]: https://github.com/ruby/ruby/blob/v2_4_2/ext/openssl/extconf.rb#L38
[2]: http://ruby-doc.org/stdlib-2.4.1/libdoc/mkmf/rdoc/MakeMakefile.html#method-i-pkg_config

### Steps to reproduce

```
$ PKG_CONFIG_PATH=/usr/lib/openssl-1.0/pkgconfig gem install puma -v 3.6.0
```

### Expected behavior

Gem installs successfully

### Actual behavior

```
[yuri@yuri-nb ~] gem install puma -v 3.6.0; n
Building native extensions.  This could take a while...
ERROR:  Error installing puma:
        ERROR: Failed to build gem native extension.

    current directory: /home/yuri/.gem/ruby/2.3.4/gems/puma-3.6.0/ext/puma_http11
/home/yuri/.rubies/ruby-2.3.4/bin/ruby -r ./siteconf20170914-19829-17mwin9.rb extconf.rb
checking for BIO_read() in -lcrypto... yes
checking for SSL_CTX_new() in -lssl... yes
checking for openssl/bio.h... yes
creating Makefile

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /home/yuri/.gem/ruby/2.3.4/extensions/x86_64-linux/2.3.0-static/puma-3.6.0/mkmf.log

current directory: /home/yuri/.gem/ruby/2.3.4/gems/puma-3.6.0/ext/puma_http11
make "DESTDIR=" clean

current directory: /home/yuri/.gem/ruby/2.3.4/gems/puma-3.6.0/ext/puma_http11
make "DESTDIR="
compiling io_buffer.c
compiling mini_ssl.c
mini_ssl.c: In function ‘get_dh1024’:
mini_ssl.c:90:5: error: dereferencing pointer to incomplete type ‘DH {aka struct dh_st}’
   dh->p = BN_bin2bn(dh1024_p, sizeof(dh1024_p), NULL);
     ^
mini_ssl.c: In function ‘engine_init_server’:
mini_ssl.c:139:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   ID sym_cert = rb_intern("cert");
   ^
mini_ssl.c:144:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   ID sym_ca = rb_intern("ca");
   ^
mini_ssl.c:166:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   DH *dh = get_dh1024();
   ^
mini_ssl.c:170:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   EC_KEY *ecdh = EC_KEY_new_by_curve_name(NID_secp521r1);
   ^
mini_ssl.c: In function ‘engine_init_client’:
mini_ssl.c:197:3: warning: ‘DTLSv1_method’ is deprecated [-Wdeprecated-declarations]
   conn->ctx = SSL_CTX_new(DTLSv1_method());
   ^
In file included from /usr/include/openssl/ct.h:13:0,
                 from /usr/include/openssl/ssl.h:61,
                 from mini_ssl.c:15:
/usr/include/openssl/ssl.h:1630:1: note: declared here
 DEPRECATEDIN_1_1_0(__owur const SSL_METHOD *DTLSv1_method(void)) /* DTLSv1.0 */
 ^
make: *** [Makefile:239: mini_ssl.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/yuri/.gem/ruby/2.3.4/gems/puma-3.6.0 for inspection.
Results logged to /home/yuri/.gem/ruby/2.3.4/extensions/x86_64-linux/2.3.0-static/puma-3.6.0/gem_make.out
```

### System configuration

**Ruby version**: 2.3.4p301
**Puma version**: 3.6.0

```
current directory: /home/jesse/.asdf/installs/ruby/2.3.1/lib/ruby/gems/2.3.0/gems/puma-3.3.0/ext/puma_http11
/home/jesse/.asdf/installs/ruby/2.3.1/bin/ruby -r ./siteconf20180706-736-l02lxz.rb extconf.rb
checking for BIO_read() in -lcrypto... yes
checking for SSL_CTX_new() in -lssl... yes
checking for openssl/bio.h... yes
creating Makefile

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /home/jesse/.asdf/installs/ruby/2.3.1/lib/ruby/gems/2.3.0/extensions/x86_64-linux/2.3.0-static/puma-3.3.0/mkmf.log

current directory: /home/jesse/.asdf/installs/ruby/2.3.1/lib/ruby/gems/2.3.0/gems/puma-3.3.0/ext/puma_http11
make "DESTDIR=" clean

current directory: /home/jesse/.asdf/installs/ruby/2.3.1/lib/ruby/gems/2.3.0/gems/puma-3.3.0/ext/puma_http11
make "DESTDIR="
compiling mini_ssl.c
mini_ssl.c: In function ‘get_dh1024’:
mini_ssl.c:90:5: error: dereferencing pointer to incomplete type ‘DH {aka struct dh_st}’
   dh->p = BN_bin2bn(dh1024_p, sizeof(dh1024_p), NULL);
     ^~
mini_ssl.c: In function ‘engine_init_server’:
mini_ssl.c:161:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   DH *dh = get_dh1024();
   ^~
mini_ssl.c:165:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   EC_KEY *ecdh = EC_KEY_new_by_curve_name(NID_secp521r1);
   ^~~~~~
mini_ssl.c: In function ‘engine_init_client’:
mini_ssl.c:192:3: warning: ‘DTLSv1_method’ is deprecated [-Wdeprecated-declarations]
   conn->ctx = SSL_CTX_new(DTLSv1_method());
   ^~~~
In file included from /usr/include/openssl/ct.h:13:0,
                 from /usr/include/openssl/ssl.h:61,
                 from mini_ssl.c:15:
/usr/include/openssl/ssl.h:1645:1: note: declared here
 DEPRECATEDIN_1_1_0(__owur const SSL_METHOD *DTLSv1_method(void)) /* DTLSv1.0 */
 ^
make: *** [Makefile:239: mini_ssl.o] Error 1

make failed, exit code 2
```
This would indeed be handy. I'm trying to install Puma in Ruby 2.3.1 with LibreSSL (or OpenSSL 1.0.1f but I can't get Ruby to install correctly, anyway)

--- a/ext/puma_http11/extconf.rb
+++ b/ext/puma_http11/extconf.rb
@@ -11,9 +11,18 @@ end
 unless ENV["DISABLE_SSL"]
   dir_config("openssl")
 
-  if %w'crypto libeay32'.find {|crypto| have_library(crypto, 'BIO_read')} and
+  found_ssl = if pkg_config 'openssl'
+    puts 'using OpenSSL pkgconfig (openssl.pc)'
+    true
+  elsif %w'crypto libeay32'.find {|crypto| have_library(crypto, 'BIO_read')} &&
       %w'ssl ssleay32'.find {|ssl| have_library(ssl, 'SSL_CTX_new')}
+    true
+  else
+    puts '** Puma will be compiled without SSL support'
+    false
+  end
 
+  if found_ssl
     have_header "openssl/bio.h"
 
     # below is  yes for 1.0.2 & later

